File: CombinationSum/CombinationSumv1.java
import java.util.*;

public class CombinationSumv1 {
    public static void main(String[] args) {
        int[] candidates = {2, 3, 6, 7};
        int target = 7;
        System.out.println(combinationSum(candidates, target));
    }

    public static List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(candidates);
        combinationSumHelper(candidates, target, 0, new ArrayList<>(), result);
        return result;
    }

    private static void combinationSumHelper(int[] candidates, int target, int start, List<Integer> current, List<List<Integer>> result) {
        if (target == 0) {
            result.add(new ArrayList<>(current));
            return;
        }
        for (int i = start; i < candidates.length; i++) {
            if (candidates[i] > target) break;
            current.add(candidates[i]);
            combinationSumHelper(candidates, target - candidates[i], i, current, result);
            current.remove(current.size() - 1);
        }
    }
}

File: 39. CombinationSum/CombinationSumv1.py
def combinationSum(candidates, target):
    result = []
    combinationSumHelper(candidates, target, 0, [], result)
    return result

def combinationSumHelper(candidates, remaining, start, current, result):
    if remaining == 0:
        result.append(list(current))
        return

    for i in range(start, len(candidates)):
        if candidates[i] > remaining:
            continue  # skip if number too large
        current.append(candidates[i])
        combinationSumHelper(candidates, remaining - candidates[i], i, current, result)
        current.pop()

if __name__ == "__main__":
    candidates = [2, 3, 6, 7]
    target = 7
    print(combinationSum(candidates, target))
